cilium/lb/main.go

```diff
@@ -140,240 +152,242 @@ func init() {
+	key := parseServiceKey(ctx.Args().Get(0))
 
-	address := ctx.Args().Get(firstArg + 2)
-	target := net.ParseIP(address)
-	if target == nil {
-		fmt.Fprintf(os.Stderr, "Unable to parse address: %s\n", address)
+	svc, err := lbmap.LookupService(key)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%s\n", err)
 		os.Exit(1)
 	}
 
-	if ipv6 {
-		if target.To4() != nil {
-			fmt.Fprintf(os.Stderr, "Expecting an IPv6 address, got: %s\n", address)
-			os.Exit(1)
-		}
-
-		return lbmap.NewService6Value(parseUint16(ctx, firstArg), target,
-			parseUint16(ctx, firstArg+3), parseUint16(ctx, firstArg+1))
-	} else {
-		if target.To4() == nil {
-			fmt.Fprintf(os.Stderr, "Expecting an IPv4 address, got: %s\n", address)
-			os.Exit(1)
-		}
+	fmt.Printf("%s = %s\n", key.String(), svc.String())
+}
 
-		return lbmap.NewService4Value(parseUint16(ctx, firstArg), target,
-			parseUint16(ctx, firstArg+3), parseUint16(ctx, firstArg+1))
+func cliLookupRevNat(ctx *cli.Context) {
+	key := parseRevNatKey(ctx.Args().Get(0))
+	val, err := lbmap.LookupRevNat(key)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%s\n", err)
+		os.Exit(1)
 	}
+
+	fmt.Printf("%v = %v\n", key, val)
 }
 
-func parseRevNat(ctx *cli.Context, ipv6 bool, firstArg int) lbmap.RevNatValue {
-	if len(ctx.Args()) < (firstArg + 2) {
-		printUsageAndExit(ctx)
+func cliUpdateService(ctx *cli.Context) {
+	key := parseServiceKey(ctx.String("frontend"))
+	svc := key.NewValue().(lbmap.ServiceValue)
+	backends := []*net.TCPAddr{}
+
+	proto := "tcp4"
+	if key.IsIPv6() {
+		proto = "tcp6"
 	}
 
-	address := ctx.Args().Get(firstArg)
-	ip := net.ParseIP(address)
-	if ip == nil {
-		fmt.Fprintf(os.Stderr, "Unable to parse address: %s\n", address)
-		os.Exit(1)
+	revNat := ctx.Int("id")
+
+	backendList := ctx.StringSlice("backend")
+	if len(backendList) == 0 {
+		fmt.Printf("Reading backend list from stdin...\n")
+
+		scanner := bufio.NewScanner(os.Stdin)
+
+		for scanner.Scan() {
+			backendList = append(backendList, scanner.Text())
+		}
 	}
 
-	if ipv6 {
-		if ip.To4() != nil {
-			fmt.Fprintf(os.Stderr, "Expecting an IPv6 address, got: %s\n", address)
+	for k := range backendList {
+		tcpAddr, err := net.ResolveTCPAddr(proto, backendList[k])
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "%s\n", err)
+			os.Exit(1)
+		}
+
+		if tcpAddr.IP.To4() != nil && key.IsIPv6() {
+			fmt.Fprintf(os.Stderr, "Address mismatch between frontend and backend %s\n",
+				backendList[k])
 			os.Exit(1)
 		}
 
-		return lbmap.NewRevNat6Value(ip, parseUint16(ctx, 2))
-	} else {
-		if ip.To4() == nil {
-			fmt.Fprintf(os.Stderr, "Expecting an IPv4 address, got: %s\n", address)
+		if key.GetPort() == 0 && tcpAddr.Port != 0 {
+			fmt.Fprintf(os.Stderr, "L4 backend found (%v) with L3 frontend\n", tcpAddr)
 			os.Exit(1)
 		}
 
-		return lbmap.NewRevNat4Value(ip, parseUint16(ctx, 2))
+		backends = append(backends, tcpAddr)
 	}
-}
 
-func cliLookupService(ctx *cli.Context) {
-	key := parseServiceKey(ctx, 0)
+	idx := 1
```

Comment:
I don't want to do this unconditionally and make the entire CLI require a consul instance and everything. A better way would be to provide a REST API through the daemon instead.